src/pkg/p2p/preheat/job.go

```diff
@@ -0,0 +1,225 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package preheat
+
+import (
+	"time"
+
+	"github.com/goharbor/harbor/src/jobservice/job"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/pkg/p2p/preheat/models/provider"
+	pr "github.com/goharbor/harbor/src/pkg/p2p/preheat/provider"
+	"github.com/goharbor/harbor/src/pkg/p2p/preheat/provider/auth"
+)
+
+const (
+	// parameter keeps the preheating provider instance info.
+	preheatParamProvider = "provider"
+	// parameter keeps the preheating artifact (image) info.
+	preheatParamImage = "image"
+	// checkInterval indicates the interval of loop check.
+	checkInterval = 10 * time.Second
+	// checkTimeout indicates the overall timeout of the loop check.
+	checkTimeout = 1801 * time.Second
```

Comment:
I'm confused as to why checkTimeout use 1801, why add one seconds?

Reply:
avoid the two time-signals coming at the same time. I remembered if they come at the same time, a random selection may happen. @chlins 

select{
  case <-tk.C:
  case <-tm.C:
}